src/models/document/base-document-content.ts

```diff
@@ -547,6 +564,42 @@ export const BaseDocumentContentModel = types
     }
   }))
   .actions(self => ({
+    applyCopySpecs(copySpecs: ICopySpec[]) {
+      const targetRowMap = new Map<string, TileRowModelType>();
+
+      copySpecs.forEach(({tile, rowId, sectionId}) => {
+        // find or create the row to copy into
+        let targetRow = targetRowMap.get(rowId);
+        let insertedRowIndex = self.defaultInsertRow;
+        const insertingRow = !targetRow;
+
+        if (insertingRow) {
+          targetRow = TileRowModel.create({ sectionId });
+          if (sectionId) {
+            const sectionRows = self.getRowsInSection(sectionId);
+            if (sectionRows.length > 0) {
+              const lastRow = sectionRows[sectionRows.length - 1];
+              insertedRowIndex = self.getRowIndex(lastRow.id);
+            }
+          }
+          self.insertRow(targetRow, insertedRowIndex);
+        }
+
+        // copy the tile
+        if (targetRow) {
+          targetRowMap.set(rowId, targetRow);
```

Comment:
Could be moved into `if (insertingRow)`, I believe.

Reply:
Fixed in https://github.com/concord-consortium/collaborative-learning/pull/2462/commits/d641d0a4d2c2fcd3e17bcd3c5cafb65b53402d48.